### ISSUE TITLE
Chicken-and-egg issue with git installation/upgrade

### ISSUE THREAD

**Issue by julian45** [Linuxbrew/homebrew-core#7726]
- [x] are reporting a bug others will be able to reproduce and not asking a question. If you're not sure or want to ask a question do so on our Discourse: https://discourse.brew.sh
- [x] have a problem with `brew install` (or `upgrade`, `reinstall`) a single, official formula (not cask)? If it's a general `brew` problem please file this issue at Homebrew/brew: https://github.com/Homebrew/brew/issues/new/choose. If it's a `brew cask` problem please file this issue at https://github.com/caskroom/homebrew-cask/issues/new/choose. If it's a tap (e.g. Homebrew/homebrew-php) problem please file this issue at the tap.
- [x] ran `brew update` and can still reproduce the problem?
- [x] ran `brew doctor`, fixed all issues and can still reproduce the problem?
- [x] ran `brew gist-logs <formula>` (where `<formula>` is the name of the formula that failed) and included the output link?
- [x] if `brew gist-logs` didn't work: ran `brew config` and `brew doctor` and included their output with your issue?

Output of `brew config`:
```
HOMEBREW_VERSION: 1.6.3
ORIGIN: https://github.com/Linuxbrew/brew
HEAD: b67917c837f1fb9c5bf2494cfd89b4eee6802802
Last commit: 11 days ago
Core tap ORIGIN: https://github.com/Linuxbrew/homebrew-core
Core tap HEAD: a58ec60cde13604660a765589f64d3dcf0a1cc47
Core tap last commit: 2 hours ago
HOMEBREW_PREFIX: /home/linuxbrew/.linuxbrew
HOMEBREW_REPOSITORY: /home/linuxbrew/.linuxbrew/Homebrew
HOMEBREW_CELLAR: /home/linuxbrew/.linuxbrew/Cellar
HOMEBREW_CACHE: /home/julian/.cache/Homebrew
CPU: octa-core 64-bit 0x69e
Homebrew Ruby: 2.3.3 => /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/2.3.3_2/bin/ruby
Clang: N/A
Git: 2.16.2 => /home/linuxbrew/.linuxbrew/bin/git
Curl: 7.47.0 => /usr/bin/curl
Java: N/A
Kernel: Linux 4.4.0-17134-Microsoft x86_64 GNU/Linux
OS: Ubuntu 16.04.4 LTS (xenial)
Host glibc: 2.23
/usr/bin/gcc: 5.4.0
glibc: N/A
gcc: 5.5.0_4
xorg: 20170115_1
```
Output of `brew doctor`: `Your system is ready to brew.`

**NOTE: If it's not clear from `brew config`'s output, I'm running this on WSL.**

**What I was trying to do and why:** `brew upgrade git` to get to the most recent version
**What happened:** Running the above command produced the following output:
```
==> Upgrading 1 outdated package, with result:
git 2.16.2 -> 2.17.0
==> Upgrading git
You must: brew install git
Error: Failure while executing: git config --replace-all homebrew.private true
```
When I ran `brew install git`, as the error said, I got the following (expected) error:
```
Error: git 2.16.2 is already installed
To upgrade to 2.17.0, run `brew upgrade git`
```
For reference, here's the output of `which git`: `/home/linuxbrew/.linuxbrew/bin/git`

**What I expected to happen:** A successful upgrade of `git`
**How to reproduce:** `brew upgrade git`

It's also worth noting that this same situation occurred on a MacBook Air with normal Homebrew. My apologies if there's an issue out for this already, in either upstream Homebrew or this repo; GitHub's issue filters are super broad, so any mention of git was included in the results of my (specific) searches.

**Comment by julian45** [comment]
Since that command doesn't include `--global` or specify where to run it, I get this when running it in `~`:
`fatal: not in a git directory`

Is there a command, like a more verbose `brew upgrade`, that could show where it was running `git config --replace-all homebrew.private true` and failed?

**Comment by iMichka** [comment]
I think this command is being called from here:
https://github.com/Linuxbrew/brew/blob/master/Library/Homebrew/tap.rb#L687-L694

You should execute that from `/home/linuxbrew/.linuxbrew/Homebrew`.
Can you give the output of
`cat /home/linuxbrew/.linuxbrew/Homebrew/.git/config` before testing the `git config` command?

**Comment by julian45** [comment]
```
[core]
        repositoryformatversion = 0
        filemode = true
        bare = false
        logallrefupdates = true
        autocrlf = false
[remote "origin"]
        url = https://github.com/Linuxbrew/brew
        fetch = +refs/heads/*:refs/remotes/origin/*
[homebrew]
        analyticsmessage = true
        analyticsuuid = 4d39e066-b2e8-4898-8936-95b971ff6d6c
```

**Comment by julian45** [comment]
Just tried running `brew upgrade git` after using the `git config` command on the repo you mentioned. Same error as before. Here's the output:
```
Updating Homebrew...
==> Auto-updated Homebrew!
Updated 1 tap (homebrew/core).
==> New Formulae
brew-php-switcher
==> Updated Formulae
vim ✔                     e2fsprogs                 gst-python                restview
ammonite-repl             emscripten                gst-rtsp-server           snakemake
ansible                   fork-cleaner              gst-validate              spades
apibuilder-cli            fq                        gstreamer                 swiftformat
binaryen                  gnome-builder             keepassc                  uhd
cglm                      gst-editing-services      launch4j                  xonsh
cmake                     gst-libav                 libidn2                   youtube-dl
dash                      gst-plugins-bad           librealsense              zopfli
diff-pdf                  gst-plugins-base          mariadb
dlib                      gst-plugins-good          neofetch
docfx                     gst-plugins-ugly          pari

==> Upgrading 1 outdated package, with result:
git 2.16.2 -> 2.17.0
==> Upgrading git
You must: brew install git
Error: Failure while executing: git config --replace-all homebrew.private true
```

**Comment by iMichka** [comment]
I edited your output to remove the analytics uuid. Not super important, but it is better to keep your analytics data anonymous.

Can you now try to run `git config --replace-all homebrew.private true` from the `/home/linuxbrew/.linuxbrew/Homebrew` folder?

I am wondering what that flag is and why it is being set.

**Comment by julian45** [comment]
I tried running `brew upgrade git` again, with that flag already set, but it *again* fails with the same error. `cat`ing the config file again reveals that the flag is already set!

Again, is there any way to make `brew upgrade` more verbose, so we can see how/why it's making the error?

**Comment by julian45** [comment]
I tried running `brew upgrade git --verbose`, just to see if it did anything useful. The output was rather long, so have a look [here](https://hastebin.com/guviyamove.coffeescript) at it if you wish.

**Comment by iMichka** [comment]
Unlucky this does not seem to help.

Maybe you could put a `raise` here:
https://github.com/Linuxbrew/brew/blob/master/Library/Homebrew/tap.rb#L690

So we could get a full stack trace of the path that is used in the code and see who is calling this.

**Comment by julian45** [comment]
Just committed that change locally and ran `brew upgrade git --verbose`. According to diffchecker.com, no difference between pre-`raise` output and post except for the "Updating Homebrew..." text at the top.

**Comment by julian45** [comment]
Just checking in after a week of nothing on this issue... is there really nothing that can be done?

**Comment by iMichka** [comment]
Not sure I can help further, running out of ideas. And as I can not reproduce this ...

Did you run `git config --replace-all homebrew.private true` from the `/home/linuxbrew/.linuxbrew/Homebrew` folder? It is weird that this command works and gives no error message, but the same command fails within brew.

**Comment by julian45** [comment]
Just ran it again this morning... Now we have something that looks more informative!
```
julian ~ $ brew upgrade git
Updating Homebrew...
Error: 784: unexpected token at '{
  "adobe-air-sdk": "homebrew/cask",
  "android-ndk": "homebrew/cask",
  "android-platform-tools": "homebrew/cask",
  "android-sdk": "homebrew/cask",
  "app-engine-go-32": "homebrew/cask/google-cloud-sdk",
  "app-engine-go-64": "homebrew/cask/google-cloud-sdk",
  "auctex": "homebrew/tex",
  "avidemux": "homebrew/cask",
  "awsenv": "Luzifer/tools",
  "chromedriver": "homebrew/cask",
  "cloudfoundry-cli": "pivotal/tap",
  "cmucl": "homebrew/cask",
  "crystax-ndk": "homebrew/cask",
  "corelocationcli": "homebrew/cask",
  "dart": "dart-lang/dart",
  "dbt": "fishtown-analytics/dbt",
  "dwarf-fortress": "homebrew/cask",
  "geany": "homebrew/cask",
  "gimp": "homebrew/cask",
  "grads": "homebrew/cask",
  "gtkwave": "homebrew/cask",
  "helios": "spotify/public",
  "hexchat": "homebrew/cask",
  "horndis": "homebrew/cask",
  "inkscape": "homebrew/cask",
  "jsl": "homebrew/cask",
  "kdiff3": "homebrew/cask",
  "keybase": "homebrew/cask",
  "magit": "dunn/emacs",
  "meld": "homebrew/cask",
  "osxfuse": "homebrew/cask",
  "pdf-tools": "dunn/emacs",
  "pebble-sdk": "pebble/pebble-sdk",
  "quassel": "homebrew/cask",
  "sqlitebrowser": "homebrew/cask",
  "transmission-remote-gtk": "homebrew/cask/transmission-remote-gui",
  "tuntap": "homebrew/cask"
  "iotop": "linuxbrew/extra",
  "libpipeline": "linuxbrew/extra",
  "man-db": "linuxbrew/extra",
  "strace": "linuxbrew/extra",
}
'
Please report this bug:
  https://github.com/Linuxbrew/brew/wiki/troubleshooting
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/2.3.3_2/lib/ruby/2.3.0/json/common.rb:156:in `parse'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/2.3.3_2/lib/ruby/2.3.0/json/common.rb:156:in `parse'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/tap.rb:513:in `tap_migrations'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/tap.rb:655:in `tap_migrations'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/cmd/update-report.rb:366:in `block in report'/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/cmd/update-report.rb:345:in `each_line'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/cmd/update-report.rb:345:in `report'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/cmd/update-report.rb:565:in `add'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/cmd/update-report.rb:76:in `block in update_report'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/tap.rb:531:in `block (2 levels) in each'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/tap.rb:530:in `each'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/tap.rb:530:in `block in each'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/tap.rb:529:in `each'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/tap.rb:529:in `each'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/cmd/update-report.rb:66:in `update_report'
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/brew.rb:101:in `<main>'
==> Upgrading 1 outdated package, with result:
git 2.16.2 -> 2.17.0
==> Upgrading git
You must: brew install git
Error: Failure while executing: git config --replace-all homebrew.private true
```

**Comment by iMichka** [comment]
The above error is unrelated and has been fixed.

Can you give me the output of:
`/home/linuxbrew/.linuxbrew/Homebrew && git config --replace-all homebrew.private true`

I maybe misunderstood but it seems you did not give much details about that. At least it is unclear to me what is happening here.

**Comment by julian45** [comment]
I did that and all I got was a directory change to `/home/linuxbrew/.linuxbrew/Homebrew` and exit code `0`, i.e. a new prompt in that location.

For clarity:
```
julian ~ $ /home/linuxbrew/.linuxbrew/Homebrew && git config --replace-all homebrew.private true
julian /home/linuxbrew/.linuxbrew/Homebrew stable $ echo $?
0
julian /home/linuxbrew/.linuxbrew/Homebrew stable $
```

**Comment by jwhite007** [comment]
@julian45 see my fix at #8230
